core/src/main/java/org/infinispan/persistence/file/SingleFileStore.java

```diff
@@ -1071,53 +1298,107 @@ private void mergeAndLogEntry(FileEntry entry, int mergeCounter) {
       }
    }
+
    @Override
-   public void purge(Executor threadPool, final PurgeListener task) {
+   public Publisher<MarshallableEntry<K, V>> purgeExpired() {
       long now = timeService.wallClockTime();
-      List<KeyValuePair<Object, FileEntry>> entriesToPurge = new ArrayList<>();
-      synchronized (entries) {
-         for (Iterator<Map.Entry<K, FileEntry>> it = entries.entrySet().iterator(); it.hasNext(); ) {
-            Map.Entry<K, FileEntry> next = it.next();
-            FileEntry fe = next.getValue();
-            if (fe.isExpired(now)) {
-               it.remove();
-               entriesToPurge.add(new KeyValuePair<>(next.getKey(), fe));
-            }
-         }
-      }
+      UnicastProcessor<MarshallableEntry<K, V>> processor = UnicastProcessor.create();
+      blockingManager.runBlocking(() -> {
+         blockingPurgeExpired(now, processor);
+      }, "sfs-purgeExpired");
+      return Flowable.fromPublisher(processor.serialize());
+   }
 
+   private void blockingPurgeExpired(long now, UnicastProcessor<MarshallableEntry<K, V>> processor) {
       long stamp = resizeLock.readLock();
       try {
-         for (Iterator<KeyValuePair<Object, FileEntry>> it = entriesToPurge.iterator(); it.hasNext(); ) {
-            KeyValuePair<Object, FileEntry> next = it.next();
-            FileEntry fe = next.getValue();
-            if (fe.isExpired(now)) {
-               it.remove();
-               try {
-                  free(fe);
-               } catch (Exception e) {
-                  throw new PersistenceException(e);
+         for (Map<K, FileEntry> segmentEntries : entries) {
+            if (segmentEntries == null)
+               continue;
+
+            List<KeyValuePair<Object, FileEntry>> entriesToPurge = new ArrayList<>();
+            synchronized (segmentEntries) {
+               for (Iterator<Map.Entry<K, FileEntry>> it = segmentEntries.entrySet().iterator(); it.hasNext(); ) {
+                  Map.Entry<K, FileEntry> next = it.next();
+                  FileEntry fe = next.getValue();
+                  if (fe.isExpired(now)) {
+                     it.remove();
+                     entriesToPurge.add(new KeyValuePair<>(next.getKey(), fe));
+                  }
                }
-               if (task != null) task.entryPurged(next.getKey());
             }
-         }
 
-         // Disk space optimizations
-         synchronized (freeList) {
-           processFreeEntries();
+            for (ListIterator<KeyValuePair<Object, FileEntry>> it = entriesToPurge.listIterator(); it.hasNext(); ) {
+               KeyValuePair<Object, FileEntry> next = it.next();
+               FileEntry fe = next.getValue();
+               if (fe.isExpired(now)) {
+                  it.set(null);
+
+                  // Already "locked" by being removed from the segmentEntries map
+                  MarshallableEntry<K, V> entry = readFromDisk(fe, next.getKey(), false, false);
```

Comment:
```suggestion
                  MarshallableEntry<K, V> entry = readFromDisk(fe, next.getKey(), true, true);
```
The goal is that expiration will have values moving forward. This makes me wonder if we should have had an argument to ask for the value on `purgeExpired`.

Reply:
Yeah, a parameter (to be set to true only when an expiration listener has been registered) sounds good.
I thought the value and metadata weren't required because both ClusterExpirationManager and ExpirationManagerImpl expect them to be missing, e.g.
https://github.com/infinispan/infinispan/blob/a40d46817d5f41c74f0f96145d23a8e1e17c6004/core/src/main/java/org/infinispan/expiration/impl/ClusterExpirationManager.java#L439

Reply:
That is just the way it worked before, the new SPI was trying to change that so the values would be returned. That way we could finally make Paul happy :)

Reply:
After spending a few good hours debugging a failure in `ExpirationSingleFileStoreDistListenerFunctionalTest.testExpirationOfStoreWhenDataNotInMemory` I finally remembered why I didn't read the value and metadata :)

https://github.com/danberindei/infinispan/blob/edcc01b34c09d991c9b447b47d2d97acfc556e99/core/src/main/java/org/infinispan/interceptors/impl/CallInterceptor.java#L560

The comment is not completely correct: even on the node where purgeExpired did not remove the entry yet, RemoveExpiredCommand has flag SKIP_CACHE_LOAD set, so it can only find the entry in the data container. If the entry does not exist in the data container, the RemoveExpiredCommand fails, and the expiration listener is not invoked, failing the test.

I'll add a comment in the store as well, if I forgot it once I'll surely forget it again!

Reply:
@wburns I added the comment, let me know if you had another fix in mind.